### DISCUSSION TITLE
conftest fixtures dependent on partially configured plugin

### DISCUSSION THREAD

**Discussion by couteau** [pytest-dev/pytest#12758] [Q&A] [answered]
This issue may be too specific to be considered a bug and may not warrant a feature request, so I'm posting here first to see if anyone has a good solution. 

I use `conftest` modules, both at the top level and in sub folders of my test folder, to define a number of fixtures that are used throughout my tests. Some of these fixtures are dependent on a plugin (specifically here, the `pytest_qgis` plugin). At the time my `conftest` modules are imported/executed, the plugin is not fully configured -- specifically, the `pytest_configure` hook has not been run. 

In my particular circumstance, this creates a problem, because the fixtures I am creating in my `conftest` files depend on python classes/functions that the plugin configures and/or mocks, and even importing those classes/functions before the configuration/mocks are set up causes unwanted code to execute (or to execute without being properly configured).

I understand that the `conftest` modules are loaded before the plugin configure hooks are executed is because the `conftest` modules themselves are plugins, and the hooks are executed in a particular order for all plugins. But where all I am doing in the `conftest` file is defining fixtures, loading could be deferred until after the other plugins are all fully configured.

I am currently working around the problem by putting the shared fixtures in a separate fixtures.py file instead of in conftest.py, and then importing it in each test module that needs to use those fixtures (which is most of them). This is less than ideal for a number of reasons. One possible enhancement to address this situation would be that, where shared figures need to be created by other plugin features are not required, instead of putting them in conftest.py, they could be placed in a different file (say 'features.py') that is loaded later in the test setup process, after all the plugins have been loaded and all plugin hooks executed. 

Does this seem like an enhancement that would be considered and that would have application outside of my narrow circumstance?

**Comment by RonnyPfannschmidt** [comment, marked as answer]
At first glance it sounds like there's a lot of stuff broken 

You can self implement your proposed solution by implementing a try last marked pytest- configure hook that registers the new plugin after import 


If that's not feasible it can also be done in the session start Hook 

It would be better if you could get by in a different way, it's fragile

**Comment by couteau** [comment]
Thanks! Using a `trylast` hook in `conftest` to load my plugin-dependent fixtures worked. I had previously tried just importing the module with the fixtures from a `trylast` hook, which didn't work, but putting them in a class and using the `config.pluginmanager` to register them works great.
